### ISSUE TITLE
Non-determinism in Scoring / Tuning

### ISSUE THREAD

**Issue by jason-fries** [HazyResearch/metal#100]
Currently dev checkpointing is non-deterministic, with the best model producing a different score when reloaded and evaluated at the end of a training loop. The model is also evaluated twice at the end of the loop.

```
....
[E:46]	Train Loss: 0.008	Dev score: 0.598
Saving model at iteration 47 with best score 0.612
[E:47]	Train Loss: 0.008	Dev score: 0.612
Saving model at iteration 48 with best score 0.622
[E:48]	Train Loss: 0.008	Dev score: 0.622
[E:49]	Train Loss: 0.007	Dev score: 0.593
Restoring best model from iteration 48 with score 0.622
Finished Training
Accuracy: 0.829
Precision: 0.470
Recall: 0.887
F1: 0.615
        y=1    y=2   
 l=1    55     62    
 l=2     7     279   
Tuner - using validation_metric=f1
F1: 0.612
        y=1    y=2   
 l=1    56     65    
 l=2     6     276   
```
In general, scoring is non-deterministic. Can scoring setup to be deterministic conditioned on the user provided seed, with something like:

```
# save rng state, seed for deterministic evaluation
rng_gpu = torch.cuda.random.get_rng_state_all()
rng_cpu = torch.random.get_rng_state()
torch.cuda.random.manual_seed_all(self.seed)
torch.random.manual_seed(self.seed)

# evaluate model here

# restore rng state to all devices
torch.cuda.set_rng_state_all(rng_gpu)
torch.random.set_rng_state(rng_cpu)
```

**Comment by ajratner** [comment]
@jason-fries thanks for catching this!  I'll take a look later today

**Comment by ajratner** [comment]
Hey @jason-fries I'm assuming the example above is with the GPU, and the LSTM? Do you know if it occurs without the GPU?  And could you let me know how you had the model configured?

I'm a bit behind but going to try to isolate and correct as suggested on [this branch](https://github.com/HazyResearch/metal/tree/model-determinism)

Thanks,
Alex

**Comment by jason-fries** [comment]
Just checked -- it's also non-deterministic on CPU.

**Comment by ajratner** [comment]
Thanks! And could you share the model class (LSTM?) and kwargs / config used?

**Comment by jason-fries** [comment]
Perhaps I don't follow how the `RandomSearchTuner` is configured but there are couple things that seem weird/buggy to me.

For random seeding, I'd expect:
- A user-provided seed for model weight initialization that's fixed during hyperparameter search.
- A fixed, separate seed for sampling hyperparameters.

If I pass a single seed into the model search grid (see below) and set a different fixed seed for the `RandomSearchTuner` initialization, then the same hyperparameters settings are sampled for each trained model. If you remove the seed from the search space, then every model config gets a new seed by default (which seems weird).  

The best model also seems to be chosen using accuracy not matter what is chosen as 'validation_metric'. How do you specify that the best model should be selected based on `validation_metric`?

Model config dictionary
```
{'input_module': lstm,
 'lstm_reduction': 'attention',
 'middle_modules': [Dropout(p=0.5],
 'seed': 123,
 'show_plots': False,
 'train_config': {'batch_size': 64,
                  'checkpoint': True,
                  'checkpoint_config': {'checkpoint_min': -1,
                                        'checkpoint_runway': 0},
                  'data_loader_config': {'batch_size': 16, 'num_workers': 8},
                  'disable_prog_bar': True,
                  'l2': 0.00166,
                  'n_epochs': 5,
                  'optimizer_config': {'adam_config': {'betas': (0.9,
                                                                 0.999)},
                                       'optimizer': 'adam',
                                       'optimizer_common': {'lr': 0.01},
                                       'sgd_config': {'momentum': 0.9}},
                  'print_every': 1,
                  'scheduler_config': {'exponential_config': {'gamma': 0.9},
                                       'lr_freeze': 0,
                                       'plateau_config': {'factor': 0.5,
                                                          'min_lr': 1e-05,
                                                          'patience': 1,
                                                          'threshold': 0.0001},
                                       'scheduler': 'reduce_on_plateau'},
                  'validation_freq': 1,
                  'validation_metric': 'f1'},
 'use_cuda': False,
 'verbose': True}
```

and 
```
search_space = {
        'lr': {'range': [0.001, 0.01], 'scale': 'log'},
        'seed':[args.seed]
}

searcher = RandomSearchTuner(EndModel, seed=args.seed + 1, log_writer_class=TensorBoardWriter,  **log_config)
end_model = searcher.search(search_space,
                                dev_loader,
                                train_args=[train_loader],
                                init_args=init_args,
                                init_kwargs=init_kwargs,
                                train_kwargs=model_config['train_config'],
                                max_search=args.n_model_search)
```

**Comment by ajratner** [comment]
@jason-fries looking into this now; one more question, has the issue only occurred when using the tuner, or also when just training the model with dev set checkpointing?

**Comment by ajratner** [comment]
Okay well either way, I now have a basic model determinism test that is failing just with the LSTM on CPU, so probably need to fix the seeding in that module; doing this first

**Comment by ajratner** [comment]
@jason-fries LSTM with randomly initialized embeddings just needs to be seeded for determinism; passing tests now on `model-determinism` branch. Moving on to random tuner

**Comment by ajratner** [comment]
Fixed by PR #105
